Occasional failure of tests/fsm

## Steps to reproduce the issue

Build yosys (commit a9b61080) and abc (commit 83f54185) from source.
I have a weird process for this step that I'm 99% sure is not relevant to this bug.
`cd tests/fsm && bash run-test.sh "-S 4605826776001879497"`

## Expected behavior

(as seen with a different seed)
`generating tests..`
`PRNG seed: 6301643816686424051`
`running tests..`
`[0]K[1]K[2]K[3]K[4]K[5]T[6]K[7]K[8]K[9]K[10]K[11]K[12]K[13]K[14]K[15]K[16]K[17]K[18]K[19]K[20]K[21]K[22]K[23]K[24]K[25]K[26]K[27]K[28]K[29]K[30]K[31]K[32]K[33]K[34]K[35]K[36]K[37]K[38]K[39]K[40]K[41]K[42]K[43]K[44]K[45]K[46]K[47]K[48]K[49]K`

## Actual behavior

`generating tests..`
`PRNG seed: 4605826776001879497`
`running tests..`
`[0]K[1]K[2]K[3]K[4]K[5]K[6]K[7]K[8]K[9]K[10]K[11]K[12]K[13]K[14]K[15]K[16]K[17]K[18]K[19]K[20]K[21]K[22]K[23]K[24]K[25]K[26]K[27]K[28]K[29]Warning: Regarding the user-specified fsm_encoding attribute on gate.state:`
 `  Users of state reg look like FSM recoding might result in larger circuit.`
 `   Doesn't look like a proper FSM. Possible simulation-synthesis mismatch!`
`K[30]K[31]terminate called after throwing an instance of 'std::bad_alloc'`
`  what():  std::bad_alloc`
`make: *** [temp/makefile:160: temp/uut_00031.log] Aborted`

--- a/passes/opt/opt_expr.cc
+++ b/passes/opt/opt_expr.cc
@@ -1596,6 +1596,14 @@ skip_identity:
 				log_debug("Removing low %d A and %d B bits from cell `%s' in module `%s'.\n",
 						a_zeros, b_zeros, cell->name.c_str(), module->name.c_str());
 
+				if (y_zeros >= GetSize(sig_y)) {
+					module->connect(sig_y, RTLIL::SigSpec(0, GetSize(sig_y)));
+					module->remove(cell);
+
+					did_something = true;
+					goto next_cell;
+				}
+
 				if (a_zeros) {
 					cell->setPort(ID::A, sig_a.extract_end(a_zeros));
 					cell->parameters[ID::A_WIDTH] = GetSize(sig_a) - a_zeros;

--- a/tests/opt/bug2221.ys
+++ b/tests/opt/bug2221.ys
@@ -1,0 +1,16 @@
+read_verilog <<EOT
+module test (
+        input [1:0] a,
+        input [1:0] b,
+        output [5:0] y
+);
+
+wire [5:0] aa = {a, 4'h0};
+wire [5:0] bb = {b, 4'h0};
+
+assign y = aa * bb;
+
+endmodule
+EOT
+
+equiv_opt -assert opt_expr

